packages/graphql/letter-to-santa/optionality.md

```diff
@@ -0,0 +1,1485 @@
+# Requiredness and Optionality in TypeSpec
+
+This proposal suggests four new expressions:
+
+1. A new symbol, `!`, to mark properties as _explicitly required_.
+2. A new decorator, `@required`, to mark properties as required in specific contexts.
+3. A new decorator, `@optional`, to mark properties as optional in specific contexts.
+4. A new decorator, `@defaultRequiredness`, to remove contextual requiredness behavior in specific contexts.
+
+<br>
+
+## Goals
+
+1. Give developers a way to write shared TypeSpec that is usable across "default optional" and "default required" protocols.
+2. Provide more control over requiredness/optionality based on context.
+    1. Find a general solution to [the `@patch` problem](#patch).
+3. Promote comprehension and maintainability by aligning with existing systems.
+
+<br>
+
+# Background
+
+Let's start by examining the current state of things in TypeSpec.
+
+## Requiredness
+
+Model properties in TypeSpec can either be [optional][optional-properties] or not.
+
+This determination is made by the presence or absence of the `?` symbol after the property name.
+
+```typespec
+model User {
+  name: string; // not optional
+  email?: string; // optional
+}
+```
+
+For the most part, TypeSpec emitters assume "not optional" to mean "required".
+The above TypeSpec, for instance, will be represented as the following:
+
+### OpenAPI emitter
+
+Adds all not optional properties to the [`required` array defined by JSON schema][json-schema-required].
+
+```yaml
+required:
+  - name
+```
+
+<details>
+<summary>OpenAPI emitter output</summary>
+
+```yaml
+User:
+  type: object
+  required:
+    - name
+  properties:
+    name:
+      type: string
+    email:
+      type: string
+```
+
+</details>
+
+### JSON Schema emitter
+
+Adds all not optional properties to the [`required` array defined by JSON schema][json-schema-required].
+
+```yaml
+required:
+  - name
+```
+
+<details>
+<summary>JSON Schema emitter output</summary>
+
+```yaml
+$schema: https://json-schema.org/draft/2020-12/schema
+$id: User.yaml
+type: object
+properties:
+  name:
+    type: string
+  email:
+    type: string
+required:
+  - name
+```
+
+</details>
+
+### Protobuf emitter
+
+Proto3 [does not have required fields][protobuf-required]. All fields are optional.
```

Comment:
So, this emitter's output already differs from what is specified in the TSP schema. If applied to GraphQL, we can say that by default all fields are optional and then a field with a `!` is specifically marked as required. It is like adding an extra step to the protobuf use-case where mostly all props optional, but a few of them can be required.